README.md
seven SMS provider
===================
![Elgg Logo](https://img.shields.io/badge/Elgg-2.1.x-orange.svg?style=flat-square)

## Features

 * Send SMS using seven API

### Usage

This plugin provides a handler for SMS notification delivery for the `notifications_sms` plugin.

It also supports sending SMS messages programmatically:

```php
notify_user($member->guid, 0, $subject, $message, [
		'action' => 'something',
		'object' => $entity,
        'sms' => $sms, // Custom text of the SMS notification, falls back to 'summary' and then to 'subject'
		'summary' => $summary,
        'url' => $entity->getURL(),
], ['email', 'site', 'sms']);
```

```php
elgg_send_sms('+4987654210', 'HI2U!');
```

#### Support

Need help? Feel free to [contact us](https://www.seven.io/en/company/contact/).


CHANGES.txt
### Seven SMS 1.1.0 - June 05, 2023 ###
 * Rebranding

### Seven SMS 1.1.0 - November 22, 2021 ###
 * Add German language.

### Seven SMS 1.0.0 - May 04, 2021 ###
 * Initial release.


actions/seven/test_sms.php
<?php

if (!elgg_send_sms(get_input('to'), get_input('text'))) {
	return elgg_error_response(elgg_echo('seven:sms:error'));
}

return elgg_ok_response('', elgg_echo('seven:sms:success'));


actions/sms77/test_sms.php
<?php

if (!elgg_send_sms(get_input('to'), get_input('text'))) {
	return elgg_error_response(elgg_echo('sms77:sms:error'));
}

return elgg_ok_response('', elgg_echo('sms77:sms:success'));

views/default/admin/seven/test_sms.php
<?php

echo elgg_view_form('seven/test_sms');


start.php
<?php

/**
 * seven SMS Service
 * @author seven communications GmbH & Co. KG <support@seven.io>
 * @copyright Copyright (c) 2021-2022, sms77 e.K. ; 2023-present seven communications GmbH & Co. KG
 */
@include_once __DIR__  . '/vendor/autoload.php';

elgg_register_event_handler('init', 'system', 'seven_sms_init');

/**
 * Initialize the plugin
 * @return void
 */
function seven_sms_init() {
    elgg_register_plugin_hook_handler('send', 'sms', 'seven_sms_send');

    elgg_register_plugin_hook_handler('register', 'menu:page',
        [\Seven\ElggSms\Menus::class, 'setupPageMenu']);

    elgg_register_action('seven/test_sms', __DIR__ . '/actions/seven/test_sms.php',
        'admin');
}

/**
 * Sends SMS using seven
 * @param string $hook "send"
 * @param string $type "sms"
 * @param bool $return Has the SMS been sent?
 * @param array $params Hook params
 * @return bool | void
 */
function seven_sms_send($hook, $type, $return, $params) {
    if ($return === true) {
        return; // another plugin has already sent the SMS
    }

    return elgg_send_sms(
        elgg_extract('to', $params), elgg_extract('text', $params));
}

function seven_get_setting($key) {
    return elgg_get_plugin_setting($key, 'seven_sms');
}

/**
 * Send an SMS
 * @param string $to Phone number of the recipient
 * @param string $text Text of the SMS
 * @return bool | void
 */
function elgg_send_sms($to, $text) {
    $apiKey = seven_get_setting('api_key');

    if (!$apiKey) {
        return;
    }

    $signature = seven_get_setting('signature');
    if ($signature) {
        $text .= " $signature";
    }

    try {
        (new \Sms77\Api\Client($apiKey, 'elgg'))->sms($to, $text, [
            'debug' => seven_get_setting('debug'),
            'flash' => seven_get_setting('flash'),
            'foreign_id' => seven_get_setting('foreign_id'),
            'from' => seven_get_setting('from'),
            'label' => seven_get_setting('label'),
            'no_reload' => seven_get_setting('no_reload'),
            'performance_tracking' => seven_get_setting('performance_tracking'),
        ]);

        return true;
    } catch (Exception $e) {
        error_log($e->getMessage());
    }
}


languages/de.php
<?php

return [
    'admin:seven' => 'seven',
    'admin:seven:test_sms' => 'Test-SMS senden',

    'menu:page:header:seven' => 'seven',

    'seven:sms:api_key' => 'API-Schlüssel',
    'seven:sms:api_key:help' => 'Erhältlich auf https://www.seven.io',
    'seven:sms:debug' => 'Debug',
    'seven:sms:debug:help' => 'Keine SMS werden versandt wenn aktiviert',
    'seven:sms:error' => 'SMS-Versand fehlgeschlagen',
    'seven:sms:flash' => 'Flash',
    'seven:sms:flash:help' => 'Flash-SMS werden, geräteabhängig, '
        . 'direkt im Display angezeigt und nicht gespeichert',
    'seven:sms:foreign_id' => 'Foreign ID',
    'seven:sms:foreign_id:help' =>
        'Max. 64 Zeichen bestehend aus a-z, A-Z, 0-9, .-_@',
    'seven:sms:from' => 'Absender',
	'seven:sms:from:help' =>
        'Absender - bis 11 alphanumerischen oder 16 numerischen Zeichen',
    'seven:sms:label' => 'Label',
    'seven:sms:label:help' => 'Max. 100 Zeichen bestehend aus a-z, A-Z, 0-9, .-_@',
    'seven:sms:no_reload' => 'Duplikate erlauben',
    'seven:sms:no_reload:help' => 'Erlaube Versand von Duplikaten - identischer Text '
        . 'und Empfänger innerhalb 180 Sekunden',
    'seven:sms:send' => 'SMS versenden',
    'seven:sms:performance_tracking' => 'Performance tracking',
	'seven:sms:performance_tracking:help' =>
        'Siehe https://help.seven.io/de/performance-tracking',
    'seven:sms:signature' => 'Signatur',
	'seven:sms:signature:help' => 'Text an ausgehende Nachrichten anhängen',
    'seven:sms:success' => 'SMS versandt',
    'seven:sms:text' => 'Text',
    'seven:sms:text:help' => 'Maximal 10000 Zeichen',
    'seven:sms:to' => 'Empfängernummer(n)',
    'seven:sms:to:help' => 'Rufnummer(n) getrennt per Komma eingeben',

	'seven:settings' => 'seven Einstellungen',
];


views/default/plugins/seven_sms/settings.php
<?php
$entity = elgg_extract('entity', $vars);

$seven_switch = static function ($name) use($entity) {
    echo elgg_view_field([
        '#help' => elgg_echo("seven:sms:$name:help"),
        '#label' => elgg_echo("seven:sms:$name"),
        '#type' => 'checkbox',
        'checked' => 1 === (int)$entity->$name,
        'default' => 0,
        'name' => "params[$name]",
        'switch' => true,
        'value' => 1,
    ]);
};

$seven_text = static function($name, $required = false) use($entity) {
    echo elgg_view_field([
        '#help' => elgg_echo("seven:sms:$name:help"),
        '#label' => elgg_echo("seven:sms:$name"),
        '#type' => 'text',
        'name' => "params[$name]",
        'required' => $required,
        'switch' => true,
        'value' => $entity->$name,
    ]);
};

$seven_text('api_key', true);
$seven_text('foreign_id');
$seven_text('from');
$seven_text('label');
$seven_text('signature');

$seven_switch('debug');
$seven_switch('flash');
$seven_switch('no_reload');
$seven_switch('performance_tracking');


languages/en.php
<?php

return [
    'admin:seven' => 'seven',
    'admin:seven:test_sms' => 'Send Test SMS',

    'menu:page:header:seven' => 'seven',

    'seven:sms:api_key' => 'API key',
    'seven:sms:api_key:help' => 'Get yours at https://www.seven.io',
    'seven:sms:debug' => 'Debug',
    'seven:sms:debug:help' => 'If activated no SMS will be sent or calculated',
    'seven:sms:error' => 'SMS could not be sent',
    'seven:sms:flash' => 'Flash',
    'seven:sms:flash:help' => 'Flash SMS get directly displayed in the receivers display',
    'seven:sms:foreign_id' => 'Foreign ID',
    'seven:sms:foreign_id:help' =>
        'Max. 64 chars, allowed characters: a-z, A-Z, 0-9, .-_@',
    'seven:sms:from' => 'From',
	'seven:sms:from:help' =>
        'Sender number - up to 11 alphanumeric or 16 numeric characters',
    'seven:sms:label' => 'Label',
    'seven:sms:label:help' => 'Max. 100 chars, allowed characters: a-z, A-Z, 0-9, .-_@',
    'seven:sms:no_reload' => 'No reload',
    'seven:sms:no_reload:help' =>
        'If enabled, prevents sending the same SMS (text, type and recipient alike)'
        . ' within 180 seconds',
    'seven:sms:send' => 'Send SMS',
    'seven:sms:performance_tracking' => 'Performance tracking',
	'seven:sms:performance_tracking:help' =>
        'See https://help.seven.io/en/performance-tracking',
    'seven:sms:signature' => 'Signature',
	'seven:sms:signature:help' => 'Text to append to outgoing messages',
    'seven:sms:success' => 'SMS has been sent',
    'seven:sms:text' => 'Text',
    'seven:sms:text:help' => 'Maximum of 10000 characters',
    'seven:sms:to' => 'Recipient(s)',
    'seven:sms:to:help' => 'Enter a phone number separated by comma',

	'seven:settings' => 'seven Settings',
];
